Libplanet.Net/Consensus/ConsensusContext.Event.cs

```diff
@@ -0,0 +1,90 @@
+using System;
+using Libplanet.Action;
+using Libplanet.Net.Messages;
+
+namespace Libplanet.Net.Consensus
+{
+    public partial class ConsensusContext<T>
+        where T : IAction, new()
+    {
+        /// <inheritdoc cref="Context{T}.ExceptionOccurred"/>
+        internal event EventHandler<(long Height, int Round, Exception)>? ExceptionOccurred;
+
+        /// <inheritdoc cref="Context{T}.TimeoutOccurred"/>
+        internal event EventHandler<
+                (long Height, int Round, Step Step, TimeSpan TimeSpan)>?
+            TimeoutOccurred;
+
+        /// <inheritdoc cref="Context{T}.TimeoutProcessed"/>
+        internal event EventHandler<(long Height, int Round)>? TimeoutProcessed;
+
+        /// <inheritdoc cref="Context{T}.StateChanged"/>
+        internal event EventHandler<
+                (long Height, int MessageLogSize, int Round, Step Step)>?
+            StateChanged;
+
+        /// <inheritdoc cref="Context{T}.MessageConsumed"/>
+        internal event EventHandler<(long Height, long Round, ConsensusMessage Message)>?
+            MessageConsumed;
+
+        /// <inheritdoc cref="Context{T}.MutationConsumed"/>
+        internal event EventHandler<(long Height, long Round, System.Action)>? MutationConsumed;
+
+        private void AttachEventHandlers(Context<T> context)
+        {
+            context.ExceptionOccurred += (sender, exception) =>
+            {
+                if (sender is Context<T> { } ctx)
+                {
+                    ExceptionOccurred?.Invoke(sender, (ctx.Height, ctx.Round, exception));
+                }
+            };
+
+            context.TimeoutOccurred += (sender, tuple) =>
+            {
+                if (sender is Context<T> { } ctx)
+                {
+                    TimeoutOccurred?.Invoke(
+                        sender,
+                        (ctx.Height, ctx.Round, tuple.Step, tuple.TimeSpan));
+                }
+            };
+
+            context.TimeoutProcessed += (sender, i) =>
+            {
+                if (sender is Context<T> { } ctx)
+                {
+                    TimeoutProcessed?.Invoke(sender, (ctx.Height, i));
+                }
+            };
+
+            context.StateChanged += (sender, tuple) =>
+            {
+                if (sender is Context<T> { } ctx)
+                {
+                    StateChanged?.Invoke(
+                        sender,
+                        (ctx.Height, tuple.MessageLogSize, tuple.Round, tuple.Step));
+                }
+            };
```

Comment:
```suggestion
            context.StateChanged += (sender, tuple) =>
                StateChanged?.Invoke(
                    this,
                    (context.Height, tuple.MessageLogSize, tuple.Round, tuple.Step));
```

`StateChanged`'s owner is `ConsensusContext`, not `Context`. As "wrapped" `EventHandler`s are operating at a different layer, I think `this` as a sender would be more appropriate.

Reply:
Moved `Context` data responsibility to `Context.Event`s EventHandler, and `ConsensusContext` takes it and propagates.
fb2509329705ee26db7aaaded24fd9908602a788
0ac9fdc43be4ea3914f762752420b8b51232a6d2